mantidproject/mantid
# Fix for crop workspace memory bug

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>
**Description of work.**

This PR fixes a bug in `ExtractSpectra` (called by CropWorkspace) which was increasing memory usage, under certain conditions, after cropping a workspace. In a typical `Workspace2D` with common bins (X data is the same in all histograms), there is only one copy of the X Histogram stored in memory. A pointer to this X Histogram is stored in each of the histograms in the workspace, saving significantly on memory.
	When `ExtractSpectra` was cropping a workspace in the X dimension, the workspace would lose this property as a new X Histogram was saved for every histogram in the workspace. This PR ensures that the cropped workspace is constructed in the same memory-saving way by only saving a pointer in each histogram to the new, cropped, X Histogram.

**To test:**

Run the example crop from the issue #34753

Check that the memory usage does not rise significantly as before.
(I have found that it now stays the same, not entirely sure why. Each Y and E histogram is replaced with one of a smaller size, so you'd assume that'd be less memory, but maybe something funny is going on behind the scenes.)

Check that the resulting data is as you'd expect it to be.



Fixes #34735 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
